File: ios/calorietracker/Services/HealthData/HealthSleepAnalysis.swift
import Foundation

/// Night derivation shared with Android (`docs/health-data.md` §1.3): rows are grouped by
/// wake day (`local_day`, end-attributed), the source with the most asleep time wins,
/// overlapping intervals within that source are unioned, other sources are ignored.
nonisolated enum HealthSleepAnalysis {
    typealias Interval = (start: Int64, end: Int64)

    static let asleepCodes: Set<Int> = [
        HealthSleepStage.asleepUnspecified.rawValue, HealthSleepStage.light.rawValue,
        HealthSleepStage.deep.rawValue, HealthSleepStage.rem.rawValue,
    ]

    /// Total length of the union of `intervals`, in milliseconds.
    static func unionMs(_ intervals: [Interval]) -> Int64 {
        let sorted = intervals.filter { $0.end > $0.start }.sorted { $0.start < $1.start }
        var total: Int64 = 0
        var current: Interval?
        for interval in sorted {
            if let open = current {
                if interval.start <= open.end {
                    current = (open.start, max(open.end, interval.end))
                } else {
                    total += open.end - open.start
                    current = interval
                }
            } else {
                current = interval
            }
        }
        if let open = current {
            total += open.end - open.start
        }
        return total
    }

    /// Nights for every wake day present in `rows`, oldest first.
    static func nights(rows: [HealthSampleRow], calendar: Calendar) -> [HealthSleepNight] {
        let byDay = Dictionary(grouping: rows.filter { !$0.isDeleted }, by: \.localDay)
        return byDay.keys.sorted().compactMap { day in
            night(rows: byDay[day] ?? [], nightOf: day)
        }
    }

    /// One night from rows that all share a wake day.
    static func night(rows: [HealthSampleRow], nightOf: String) -> HealthSleepNight? {
        let candidates = rows.filter { !$0.isDeleted && $0.categoryValue != HealthSleepStage.outOfBed.rawValue }
        guard !candidates.isEmpty else { return nil }
        let bySource = Dictionary(grouping: candidates, by: \.sourceID)

        func asleepMs(_ sourceRows: [HealthSampleRow]) -> Int64 {
            unionMs(sourceRows.filter { asleepCodes.contains($0.categoryValue ?? -1) }.map { ($0.startMs, $0.endMs) })
        }

        // Longest asleep time wins; ties fall back to the most rows, then a stable name order.
        guard let (source, chosen) = bySource.max(by: { lhs, rhs in
            let l = asleepMs(lhs.value), r = asleepMs(rhs.value)
            if l != r { return l < r }
            if lhs.value.count != rhs.value.count { return lhs.value.count < rhs.value.count }
            return lhs.key > rhs.key
        }) else { return nil }

        func stageMs(_ code: HealthSleepStage) -> Int64 {
            unionMs(chosen.filter { $0.categoryValue == code.rawValue }.map { ($0.startMs, $0.endMs) })
        }

        let asleep = asleepMs(chosen)
        let awake = stageMs(.awake)
        let inBedRows = chosen.filter { $0.categoryValue == HealthSleepStage.inBed.rawValue }
        let inBed: Int64
        if inBedRows.isEmpty {
            inBed = unionMs(chosen.map { ($0.startMs, $0.endMs) })
        } else {
            inBed = unionMs(inBedRows.map { ($0.startMs, $0.endMs) })
        }
        let start = chosen.map(\.startMs).min() ?? 0
        let end = chosen.map(\.endMs).max() ?? start

        return HealthSleepNight(
            nightOf: nightOf,
            startMs: start,
            endMs: end,
            inBedS: Double(inBed) / 1000,
            asleepS: Double(asleep) / 1000,
            lightS: Double(stageMs(.light)) / 1000,
            deepS: Double(stageMs(.deep)) / 1000,
            remS: Double(stageMs(.rem)) / 1000,
            awakeS: Double(awake) / 1000,
            source: source
        )
    }
}

File: android/app/src/main/java/com/ayuvo/health/data/health/HealthSleepAnalysis.kt
package com.ayuvo.health.data.health

/** Canonical sleep stage codes shared with iOS (`category_value` on `sleep` rows). */
object HealthSleepCodes {
    const val IN_BED = 0
    const val ASLEEP_UNSPECIFIED = 1
    const val AWAKE = 2
    const val LIGHT = 3
    const val DEEP = 4
    const val REM = 5
    const val OUT_OF_BED = 6

    val ASLEEP: Set<Int> = setOf(ASLEEP_UNSPECIFIED, LIGHT, DEEP, REM)

    fun label(code: Int?): String = when (code) {
        IN_BED -> "in_bed"
        ASLEEP_UNSPECIFIED -> "asleep_unspecified"
        AWAKE -> "awake"
        LIGHT -> "light"
        DEEP -> "deep"
        REM -> "rem"
        OUT_OF_BED -> "out_of_bed"
        else -> "unknown"
    }
}

data class SleepNight(
    /** Wake day, yyyy-MM-dd. */
    val nightOf: String,
    val startMs: Long,
    val endMs: Long,
    val inBedS: Double,
    val asleepS: Double,
    val lightS: Double,
    val deepS: Double,
    val remS: Double,
    val awakeS: Double,
    val sourceId: String
)

/**
 * Nights are derived at read time (never stored): per wake day pick the source with the
 * longest asleep time, union overlapping intervals within that source and ignore other
 * sources. `out_of_bed` rows never count towards in-bed or asleep time. Identical on iOS.
 */
object HealthSleepAnalysis {

    fun nights(rows: List<HealthSampleRow>): List<SleepNight> =
        rows.asSequence()
            .filter { !it.deleted }
            .groupBy { it.localDay }
            .mapNotNull { (day, dayRows) -> nightFor(day, dayRows) }
            .sortedBy { it.nightOf }

    fun nightFor(day: String, rows: List<HealthSampleRow>): SleepNight? {
        val live = rows.filter { !it.deleted && it.categoryValue != HealthSleepCodes.OUT_OF_BED }
        if (live.isEmpty()) return null
        val bySource = live.groupBy { it.sourceId }
        val chosen = bySource.maxWithOrNull(
            compareBy<Map.Entry<String, List<HealthSampleRow>>> { (_, r) -> unionSeconds(r.filter { it.categoryValue in HealthSleepCodes.ASLEEP }) }
                .thenBy { (_, r) -> unionSeconds(r) }
                .thenBy { it.key }
        ) ?: return null
        val sourceRows = chosen.value
        val asleepRows = sourceRows.filter { it.categoryValue in HealthSleepCodes.ASLEEP }
        val inBedRows = sourceRows.filter { it.categoryValue == HealthSleepCodes.IN_BED }
        val inBed = if (inBedRows.isNotEmpty()) unionSeconds(inBedRows) else unionSeconds(sourceRows)
        return SleepNight(
            nightOf = day,
            startMs = sourceRows.minOf { it.startMs },
            endMs = sourceRows.maxOf { it.endMs },
            inBedS = inBed,
            asleepS = unionSeconds(asleepRows),
            lightS = unionSeconds(sourceRows.filter { it.categoryValue == HealthSleepCodes.LIGHT }),
            deepS = unionSeconds(sourceRows.filter { it.categoryValue == HealthSleepCodes.DEEP }),
            remS = unionSeconds(sourceRows.filter { it.categoryValue == HealthSleepCodes.REM }),
            awakeS = unionSeconds(sourceRows.filter { it.categoryValue == HealthSleepCodes.AWAKE }),
            sourceId = chosen.key
        )
    }

    /** Total seconds covered by the union of the rows' `[start, end)` intervals. */
    fun unionSeconds(rows: List<HealthSampleRow>): Double {
        if (rows.isEmpty()) return 0.0
        val sorted = rows.map { it.startMs to maxOf(it.startMs, it.endMs) }.sortedBy { it.first }
        var total = 0L
        var curStart = sorted[0].first
        var curEnd = sorted[0].second
        for (i in 1 until sorted.size) {
            val (s, e) = sorted[i]
            if (s <= curEnd) {
                if (e > curEnd) curEnd = e
            } else {
                total += curEnd - curStart
                curStart = s
                curEnd = e
            }
        }
        total += curEnd - curStart
        return total / 1000.0
    }
}
